frontend/src/metabase-lib/queries/utils/dataset.ts

```diff
@@ -53,21 +54,32 @@ export function findColumnIndexForColumnSetting(
 }
 
 export function findColumnSettingIndexForColumn(
+  query: Lib.Query,
   columnSettings: TableColumnOrderSetting[],
   column: DatasetColumn,
 ) {
-  const fieldRef = fieldRefForColumn(column);
-  const normalizedFieldRef = fieldRef ? normalizeFieldRef(fieldRef) : undefined;
-  if (normalizedFieldRef == null) {
-    return columnSettings.findIndex(
-      columnSetting => columnSetting.name === column.name,
-    );
-  }
-  const index = columnSettings.findIndex(
-    columnSetting =>
-      columnSetting.fieldRef &&
-      _.isEqual(normalizedFieldRef, normalizeFieldRef(columnSetting.fieldRef)),
+  // ignore settings without fieldRef but preserve indexes
+  const items = columnSettings.reduce<
+    { fieldRef: FieldReference; settingIndex: number }[]
+  >((items, { fieldRef }, settingIndex) => {
+    if (fieldRef != null) {
+      items.push({ fieldRef, settingIndex });
+    }
+    return items;
+  }, []);
+
+  // first try to find by fieldRef
+  const stageIndex = -1;
+  const itemIndexes = Lib.findColumnIndexesFromLegacyRefs(
+    query,
+    stageIndex,
+    [column],
+    items.map(({ fieldRef }) => normalizeFieldRef(fieldRef) as FieldReference),
   );
 
-  return index;
+  const itemIndex = itemIndexes.find(index => index >= 0);
+
+  if (itemIndex != null) {
+    return items[itemIndex].settingIndex;
+  }
```

Comment:
:x: Not all paths return a value

```suggestion
  }

  return -1;
```